File: 1143-longest-common-subsequence/1143-longest-common-subsequence.swift
class Solution {
    func longestCommonSubsequence(_ text1: String, _ text2: String) -> Int {
        var l1 = text1.count
        var l2 = text2.count
        var dp: [[Int]] = Array(repeating: Array(repeating: -1, count: l2), count: l1)
        return calculate(l1-1,l2-1, Array(text1), Array(text2), &dp)
    }
    
    private func calculate(_ i: Int,_ j: Int,_ s1: [Character],_ s2: [Character],_ dp:inout [[Int]]) -> Int {
        if i<0 || j<0 { return 0 }
        
        if dp[i][j] != -1 {
            return dp[i][j]
        }
        
        if s1[i]==s2[j] { 
            dp[i][j] = 1 + calculate(i-1,j-1, s1, s2, &dp)
            return dp[i][j]
        } else {
            dp[i][j] = max(calculate(i-1,j, s1, s2, &dp), calculate(i,j-1, s1, s2, &dp))
            return dp[i][j]
        }
    }
}

File: 1143-longest-common-subsequence/1143-longest-common-subsequence.java
class Solution {
    public int longestCommonSubsequence(String text1, String text2) {
        int l1 = text1.length();
		int l2 = text2.length();
		Integer dp[][] = new Integer[l1][l2];
		return calculate(l1-1, l2-1, text1, text2, dp);
    }
    
    public int calculate(int i1, int i2, String s, String t, Integer[][] dp) {

		if ((i1 < 0) || (i2 < 0)) {
			return 0;
		}

		if (dp[i1][i2] != null) {
			return dp[i1][i2];
		}


		int count = 0;
		if(s.charAt(i1)==t.charAt(i2)) {
			count = 1 + calculate(i1-1, i2-1,s,t,dp);
		} else {
			int first = calculate(i1-1, i2,s,t,dp);
			int second = calculate(i1, i2-1,s,t,dp);
			count = Math.max(first, second);
		}
		dp[i1][i2] = count;
		return dp[i1][i2];
	}
}